data/transactions/logic/opcodes.go

```diff
@@ -507,6 +509,34 @@ func opMatchStackChange(cx *EvalContext) (deletions, additions int) {
 	return
 }
 
+// CurrentScratchChange tells if current opcode is storing value to a scratch slot,
+// and if so, it returns scratch slot id, previous value, new value to be written;
+// otherwise, it indicates the current opcode is not a scratch slot chagne.
+func (cx *EvalContext) CurrentScratchChange() (scratchSlot uint64, newValue basics.TealValue, isScratchChange bool) {
+	currentOpcodeName := opsByOpcode[cx.version][cx.program[cx.pc]].Name
```

Comment:
I wonder if the `BeforeOpcode` (and probably `AfterOpcode`) tracer methods should be expanded to pass the opcode? It seems a little dirty that:
1. Tracers have to understand `cx` in this much detail to operate on an opcode. (After all, you'd expect any consumer of these `*Opcode()` methods to care about the opcodes!)
2. This part of the code needs to go in the `logic` package in order to have access. (which we could fix by exporting `opsByOpcode`, `version`, and `program` I suppose.

Even if we make that change, I'm ok with it being a separate PR.

@jasonpaulos What do you think?

Reply:
If we pass the OpSpec by value, then we get _most_ of the protection we need. Except that an OpSpec has, inside it, reference types. So if you manipulated  `OpSpec.OpDetails.Immediates`, you'd be manipulating the underlying structure.

Reply:
I think it does make sense to make these operations easier. Ideally this code would be placed in the simulation package.

Passing in the `OpSpec` to `Before/AfterOpcode` would definitely help. I can also imagine having an `EvalContext.GetOpSpec(pc int)` function as an alternative. Right now I have no preference between the two.

You're right in that we'd also need to expose `EvalContext.program`, since this code needs to figure out immediate args as well.

Reply:
how about Before/After opcode getting the OpSpec and the OpDetail extended to include a SlotsModified logic function which returns nothing except for store/stores, where it runs this logic.

Reply:
see https://github.com/algorand/go-algorand/pull/5563/commits/eee45a41a423480ecb976de70e653937bf811b2c